### PR TITLE
H3 basic implementation

### PR DESCRIPTION
**Pull request by stammw**

### CODE REVIEW

File: quinn-h3/src/proto/mod.rs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+macro_rules! stream_types {
+    {$($name:ident = $val:expr,)*} => {
+        impl StreamType {
+            $(pub const $name: StreamType = StreamType($val);)*
+        }
+    }
+}
+
+stream_types! {
+    CONTROL = 0x00,
```
[the remark this answers is not in the document]
Comment (stammw) [reply quinn-h3/src/proto/mod.rs]: My idea here is to have the same value formatting than in IETF documents, like [here](https://quicwg.org/base-drafts/draft-ietf-quic-http.html#rfc.section.3.2.2) and [here](https://quicwg.org/base-drafts/draft-ietf-quic-qpack.html#rfc.section.8.2)

File: quinn-h3/src/frame.rs [hunk 2]
```diff
@@ -202,8 +202,8 @@ pub struct PriorityFrame {
 impl Codec for PriorityFrame {
     fn decode<B: Buf>(buf: &mut B) -> Result<Self, UnexpectedEnd> {
         let first = buf.get_u8();
-        let pt = (0b1100_0000 & first) >> 6;
-        let dt = (0b0011_0000 & first) >> 4;
+        let pt = (0b11000000 & first) >> 6;
```
[the remark this answers is not in the document]
Comment (stammw) [reply quinn-h3/src/frame.rs]: It was a bad squash I guess thank you for catching this (again)

File: quinn-h3/src/frame.rs [hunk 3]
```diff
@@ -396,24 +397,21 @@ mod tests {
 
     #[test]
     fn settings_frame_ignores_0x_a_a() {
-        let mut buf = vec![4, 16, 8, 128, 0, 250, 218];
+        let mut buf = vec![4, 16, 8,  128, 0, 250, 218];
```
[the remark this answers is not in the document]
Comment (stammw) [reply quinn-h3/src/frame.rs]: Gone.

File: quinn-h3/src/frame.rs [hunk 4]
```diff
@@ -396,24 +397,21 @@ mod tests {
...
+        assert_eq!(decoded, HttpFrame::Settings(SettingsFrame {
+            num_placeholders: 0xfada,
+            max_header_list_size: 0xfada,
+        }));
     }
 
     #[test]
     fn settings_frame_invalid_value() {
-        let mut buf = Cursor::new(&[4, 6, 0, 255, 128, 0, 250, 218]);
+       let mut buf = Cursor::new(&[4, 6, 0, 255, 128, 0, 250, 218]);
```
[the remark this answers is not in the document]
Comment (stammw) [reply quinn-h3/src/frame.rs]: Gone.

File: quinn-h3/src/frame.rs [hunk 5]
```diff
@@ -427,15 +425,12 @@ mod tests {
 
     #[test]
     fn settings_frame_ignores_unknown_id() {
-        let mut buf = Cursor::new(&[4, 10, 0xA, 128, 0, 250, 218, 6, 128, 0, 250, 218]);
+        let mut buf = Cursor::new(&[4, 10, 0xA, 128, 0, 250, 218, 6, 128, 0, 250, 218],);
```
[the remark this answers is not in the document]
Comment (stammw) [reply quinn-h3/src/frame.rs]: Gone.

File: quinn-h3/src/proto/frame.rs [hunk 6]
```diff
@@ -66,7 +65,7 @@ impl HttpFrame {
             Type::GOAWAY => Ok(HttpFrame::Goaway(payload.get_var()?)),
             Type::MAX_PUSH_ID => Ok(HttpFrame::MaxPushId(payload.get_var()?)),
             Type::DUPLICATE_PUSH => Ok(HttpFrame::DuplicatePush(payload.get_var()?)),
-            t if t.0 > 0xb && (t.0 - 0xb) % 0x1f == 0 => Ok(HttpFrame::Reserved),
+            t if t.0 > 0x21 && (t.0 - 0x21) % 0x1f == 0 => Ok(HttpFrame::Reserved),
```
[the remark this answers is not in the document]
Comment (stammw) [reply quinn-h3/src/proto/frame.rs]: It was yes, I'm pretty sure the mess comes from an auto save (compiling) during a rebase or a squash...

File: quinn-h3/src/frame.rs [hunk 7]
```diff
@@ -0,0 +1,153 @@
+use std::io;
+
+use bytes::BytesMut;
+use futures::{Async, Poll, Stream};
+use tokio_io::AsyncRead;
+
+use super::proto::frame::{self, HttpFrame};
+
+#[derive(Debug)]
+enum Error {
```
[the remark this answers is not in the document]
Comment (stammw) [reply quinn-h3/src/frame.rs]: Done.

File: quinn-h3/src/proto/mod.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (djc) [reply quinn-h3/src/proto/mod.rs]: That makes sense then!

File: quinn-h3/examples/h3.rs [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+    ca: Option<PathBuf>,
+    /// Simulate NAT rebinding after connecting
+    #[structopt(long = "rebind")]
+    rebind: bool,
+}
+
+fn main() {
+    let opt = Opt::from_args();
+    let sdecorator = slog_term::PlainSyncDecorator::new(std::io::stderr());
+    let cdecorator = slog_term::PlainSyncDecorator::new(std::io::stderr());
```
Comment (Ralith) [inline quinn-h3/examples/h3.rs]: You shouldn't construct completely separate root loggers. Instead, construct a single one, then fork it with `log.new(o!(whatever))` as necessary.

Comment (stammw) [reply quinn-h3/examples/h3.rs]: Thanks!

File: quinn-h3/src/client.rs [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+                .or(Err(Error::Internal("failed to build response")))?,
+        })
+    }
+
+    pub fn response<'a>(&'a self) -> &'a Response<()> {
+        &self.response
+    }
+
+    pub fn body(self) -> RecvBody {
+        RecvBody::with_capacity(self.recv, 10240, 1024000, self.conn.clone(), self.stream_id)
```
Comment (stammw) [inline quinn-h3/src/client.rs]: Those static values aren't to be left like this. I'm not sure where to store them, and how to let the user set them:
- They could vanish on body streaming implementation. If we decide to replace the current API with streamin only.
- use setters as member of `RecvBody` like this: 
```rust
response.set_max_capacity(n);
response.set_capacity(m);
response.body().and_then(...)
```

File: quinn-h3/src/frame.rs [hunk 10]
```diff
@@ -1,519 +1,167 @@
...
-            match identifier {
-                id if id.0 & 0x0f0f == 0x0a0a => continue,
-                SettingId::NUM_PLACEHOLDERS => {
-                    settings.num_placeholders = value;
-                }
-                SettingId::MAX_HEADER_LIST_SIZE => {
-                    settings.max_header_list_size = value;
+        return match decoded {
+            Err(frame::Error::UnexpectedEnd) => match self.buf.len() {
+                Self::READ_SIZE => Err(Error::Overflow),
```
Comment (stammw) [inline quinn-h3/src/frame.rs]: This won't play well with big `Data` frames. Either the buffer should grow, either a partial frame decoding should be implemented, to let the user of this type stream chunks somehow.

I decided to leave this as is for now, looking forward to body streaming feature for the end user.

File: quinn-h3/src/qpack/encoder.rs [hunk 11]
```diff
@@ -35,17 +35,17 @@ pub enum Error {
     UnknownPrefix,
 }
 
-pub fn encode<W: BufMut>(
+pub fn encode<W: BufMut, T: Iterator<Item = HeaderField>>(
```
Comment (djc) [inline quinn-h3/src/qpack/encoder.rs]: Maybe `IntoIterator` makes more sense here?

File: quinn-h3/src/qpack/tests.rs [hunk 12]
```diff
@@ -82,7 +82,7 @@ fn blocked_header() {
         &mut enc_table.encoder(42),
         &mut block_buf,
         &mut enc_buf,
-        &[HeaderField::new("foo", "bar")],
+        vec![HeaderField::new("foo", "bar")].into_iter(),
```
Comment (djc) [inline quinn-h3/src/qpack/tests.rs]: We should make sure that you can still pass an array or slice in.

File: quinn-h3/src/client.rs [hunk 13]
```diff
@@ -94,3 +107,97 @@ impl Future for Connecting {
...
+                }
+                SendRequestState::Sent(ref mut shut) => {
+                    try_ready!(shut.poll());
+                    self.state = match mem::replace(&mut self.recv, None) {
+                        Some(r) => SendRequestState::Receiving(r),
+                        None => return Err(Error::Internal("Invalid receive state")),
+                    }
+                }
+                SendRequestState::Receiving(ref mut frames) => match try_ready!(frames.poll()) {
+                    None => return Err(Error::peer("recieved an empty response")),
```
Comment (djc) [inline quinn-h3/src/client.rs]: `recieved` -> `received` (I think you have this typo in multiple places, also in commit messages).

File: quinn-h3/src/proto/headers.rs [hunk 14]
```diff
@@ -0,0 +1,296 @@
...
+pseudo_type![
+    (METHOD, b":method"),
+    (SCHEME, b":scheme"),
+    (AUTHORITY, b":authority"),
+    (PATH, b":path"),
+    (STATUS, b":status"),
+];
+
+#[derive(Debug, Default, Eq, PartialEq)]
+pub struct Pseudo {
```
Comment (djc) [inline quinn-h3/src/proto/headers.rs]: This whole term of the "pseudo-header" isn't something I'm familiar with. Maybe add a few lines of documentation to clarify what/why it's here?

File: quinn-h3/src/proto/headers.rs [hunk 15]
```diff
@@ -197,10 +219,10 @@ pseudo_type![
 #[derive(Debug, Default, Eq, PartialEq)]
 pub struct Pseudo {
     // Request
-    pub method: Option<Method>,
-    pub scheme: Option<String<Bytes>>,
-    pub authority: Option<String<Bytes>>,
-    pub path: Option<String<Bytes>>,
+    pub(self) method: Option<Method>,
```
Comment (djc) [inline quinn-h3/src/proto/headers.rs]: Not sure what `pub(self)` is or why it's needed here?

File: quinn-h3/src/client.rs (hunk 9 above)
Comment (djc) [reply quinn-h3/src/client.rs]: I think we should have two modes of operation for bodies: (a) an all-at-once mode that keeps a growable buffer internally and returns the entire response body at once to the caller, and (b) a streaming mode that has a limited fixed-size buffer whose size is configured for the `Connection` on construction.

File: quinn-h3/src/frame.rs [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+
+impl<R> Stream for FrameStream<R>
+where
+    R: AsyncRead,
+{
+    type Item = HttpFrame;
+    type Error = Error;
+
+    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        self.buf.resize(self.buf.len() + Self::READ_SIZE, 0);
```
[the remark this answers is not in the document]
Comment (stammw) [reply quinn-h3/src/frame.rs]: Right, all the non-zero `AsyncReady(n)` will trigger a realloc I guess. 

I won't be able to fix this before my holidays, so i'll mark it with a TODO and open an issue for it.

File: quinn-h3/src/lib.rs [hunk 17]
```diff
@@ -9,6 +9,16 @@ extern crate proptest;
 #[macro_use]
 extern crate assert_matches;
 
+pub mod client;
 mod frame;
 pub mod proto;
 pub mod qpack;
+pub mod server;
+
+#[derive(Clone, Default)]
```
[the remark this answers is not in the document]
Comment (stammw) [reply quinn-h3/src/lib.rs]: This is fixed in a subsequent commit with an `impl Default {}`

File: quinn-h3/src/qpack/tests.rs (hunk 12 above)
Comment (stammw) [reply quinn-h3/src/qpack/tests.rs]: if we want `encode()` to accept an `IntoIterator<HeaderField>`, aren't we obligated to pass an owned value ?

File: quinn-h3/src/client.rs (hunk 13 above)
Comment (stammw) [reply quinn-h3/src/client.rs]: Fixed in 60b761c

File: quinn-h3/src/lib.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (stammw) [reply quinn-h3/src/lib.rs]: Fixed in a subsenquent commit, with an `impl Default {}`.

File: quinn-h3/src/frame.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (djc) [reply quinn-h3/src/frame.rs]: Sounds good! When are you going on holidays?

File: quinn-h3/src/qpack/tests.rs (hunk 12 above)
Comment (djc) [reply quinn-h3/src/qpack/tests.rs]: `IntoIterator` is implemented by the std library for slices.

File: quinn-h3/src/frame.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (stammw) [reply quinn-h3/src/frame.rs]: I'll be on holidays tomorrow (28/05). I'll try to fix everything, open some issues and track them into a master issue before that.

File: quinn-h3/src/qpack/encoder.rs (hunk 11 above)
Comment (stammw) [reply quinn-h3/src/qpack/encoder.rs]: Fixed in d47c107

File: quinn-h3/src/qpack/tests.rs (hunk 12 above)
Comment (stammw) [reply quinn-h3/src/qpack/tests.rs]: Fixed in d47c107

**Comment by stammw** [comment]
All file-level `allow(dead_code)` have been removed.

File: quinn-h3/src/proto/frame.rs [hunk 18]
```diff
@@ -426,6 +426,8 @@ mod tests {
             HttpFrame::Settings(SettingsFrame {
                 num_placeholders: 0xfada,
                 max_header_list_size: 0xfada,
+                qpack_max_table_capacity: 0,
```
[the remark this answers is not in the document]
Comment (stammw) [reply quinn-h3/src/proto/frame.rs]: Sorry but this one is quite hard to rebase as it is part of a move...

File: quinn-h3/src/proto/headers.rs (hunk 14 above)
Comment (stammw) [reply quinn-h3/src/proto/headers.rs]: done in f869a8e

File: quinn-h3/src/proto/headers.rs (hunk 15 above)
Comment (stammw) [reply quinn-h3/src/proto/headers.rs]: Wasn't needed, thanks.

**Pull request merged**